addons/hr_holidays/models/hr_leave_allocation.py

```diff
@@ -469,77 +470,134 @@ def _process_accrual_plans(self, date_to=False, force_period=False, log=True):
+                    period_start, period_end = period_boundaries(current_level, allocation.lastcall)
+                    nextcall = current_level._get_next_date(allocation.nextcall)
+
+                    # If there is a carryover date coming nextcall is set to it, so the cron is called when we need to remove days
+                    carryover_date = allocation._get_carryover_date(allocation.nextcall)
+                    if allocation.nextcall < carryover_date < nextcall:
+                        nextcall = min(nextcall, carryover_date)
+                    # If it's the carry-over date, adjust days using current level's carry-over policy, then continue
+                    if allocation.nextcall == carryover_date:
+                        carryover(allocation, current_level)
+
+                    if allocation.nextcall == period_end:  # on a regular allocation date (i.e. not carryover; not level change)
+                        if allocation.accrual_plan_id.transition_mode == 'immediately':
+                            # loop over the different levels that overlap the current period and add days
+                            # as a prorata of the duration each level was in effect
+                            subperiod_start = max(period_start, first_level_start_date)
+                            while current_level_idx < len(level_ids) - 1:
+                                next_level = level_ids[current_level_idx + 1]
+                                current_level_last_date = allocation.date_from + get_timedelta(next_level.start_count, next_level.start_type)
+                                if current_level_last_date >= period_end:
+                                    break
+
+                                allocation.with_context(
+                                    override_accrual_dates=(subperiod_start, current_level_last_date)
+                                )._add_days_to_allocation(
+                                    current_level, current_level_maximum_leave, leaves_taken, period_start, period_end
+                                )
+
+                                subperiod_start = current_level_last_date
+                                current_level = next_level
+                                current_level_maximum_leave = get_max_days(allocation, current_level)
+                                current_level_idx += 1
+                            # Last iteration
+                            allocation.with_context(
+                                override_accrual_dates=(subperiod_start, period_end)
+                            )._add_days_to_allocation(
+                                current_level, current_level_maximum_leave, leaves_taken, period_start, period_end
+                            )
+                        else:
+                            allocation._add_days_to_allocation(current_level, current_level_maximum_leave, leaves_taken, period_start, period_end)
```

Comment:
At this point it start to get really hard to follow tbh maybe it could be worth to create 2 helper function for it or 

also I see that this part
```py
                    current_level, current_level_idx, current_level_maximum_leave = get_level_data(allocation, allocation.nextcall)

                    # get the period start and end for the **ongoing** period (as of `allocation.nextcall`), since we accrue days in advance
                    period_start, period_end = period_boundaries(current_level, allocation.nextcall)
                    nextcall = period_end

                    # If there is a carryover date coming, nextcall is set to it -- so the cron is called when we need to remove days
                    carryover_date = allocation._get_carryover_date(allocation.nextcall)
                    if allocation.nextcall < carryover_date < nextcall:
                        nextcall = min(nextcall, carryover_date)
                    # if it's the carry-over date, adjust days using current level's carry-over policy, then continue
                    if allocation.nextcall == carryover_date:
                        carryover(allocation, current_level)
```

is basically duplicated as well as 

```py
                     allocation.lastcall = allocation.nextcall
                    allocation.nextcall = nextcall
                    if force_period and allocation.nextcall > date_to:
                        allocation.nextcall = date_to
                        force_period = False
```
            
I'm wondering if creating a method of the structure

```py
                    current_level, current_level_idx, current_level_maximum_leave = get_level_data(allocation, allocation.nextcall)

                    # get the period start and end for the **ongoing** period (as of `allocation.nextcall`), since we accrue days in advance
                    period_start, period_end = period_boundaries(current_level, allocation.nextcall)
                    nextcall = period_end

                    # If there is a carryover date coming, nextcall is set to it -- so the cron is called when we need to remove days
                    carryover_date = allocation._get_carryover_date(allocation.nextcall)
                    if allocation.nextcall < carryover_date < nextcall:
                        nextcall = min(nextcall, carryover_date)
                    # if it's the carry-over date, adjust days using current level's carry-over policy, then continue
                    if allocation.nextcall == carryover_date:
                        carryover(allocation, current_level)

###### biffurcation call #######

                    allocation.lastcall = allocation.nextcall
                    allocation.nextcall = nextcall
                    if force_period and allocation.nextcall > date_to:
                        allocation.nextcall = date_to
                        force_period = False

```

would be better
it's an open question tho really not sure on this one I make it that way because of the number of args